logsmanagement/db_api.c

```diff
@@ -0,0 +1,1096 @@
+                      "CREATE TABLE IF NOT EXISTS " LOGS_TABLE "("
+                      "Id 					INTEGER 	PRIMARY KEY,"
+                      "FK_BLOB_Id			INTEGER		NOT NULL,"
+                      "BLOB_Offset			INTEGER		NOT NULL,"
+                      "Timestamp 			INTEGER		NOT NULL,"
+                      "Msg_compr_size 		INTEGER		NOT NULL,"
+                      "Msg_decompr_size 	INTEGER		NOT NULL,"
+                      "FOREIGN KEY (FK_BLOB_Id) REFERENCES " BLOBS_TABLE " (Id) ON DELETE CASCADE ON UPDATE CASCADE"
+                      ");",
+                      0, 0, &err_msg);
+		if (unlikely(rc != SQLITE_OK)) {
+			sqlite3_free(err_msg);
+			fatal("Failed to create %s. SQL error: %s\n", LOGS_TABLE, err_msg);
+		} 
+		// else debug(D_LOGS_MANAG, "Table %s created successfully\n", LOGS_TABLE);
+		
+		// Create index on LOGS_TABLE Timestamp
+		/* TODO: If this doesn't speed up queries, check SQLITE R*tree module. 
+		 * Requires benchmarking with/without index. */
+		rc = sqlite3_exec(p_file_infos_arr->data[i]->db,
+						  "CREATE INDEX IF NOT EXISTS logs_timestamps_idx "
+						  "ON " LOGS_TABLE "(Timestamp);",
+						  0, 0, &err_msg);
+		if (unlikely(rc != SQLITE_OK)) {
+			fatal("Failed to create logs_timestamps_idx. SQL error: %s", err_msg);
+			sqlite3_free(err_msg);
+		} 
+		// else debug(D_LOGS_MANAG, "logs_timestamps_idx created successfully\n");
+
+		/* Remove excess BLOBs beyond BLOB_MAX_FILES (from both DB and disk 
+		 * storage). This is useful if BLOB_MAX_FILES is reduced after an agent
+		 * restart (for example, if in the future it is not hardcoded, but 
+		 * instead it is read from the configuration file). LOGS_TABLE entries
+		 * should be deleted automatically (due to ON DELETE CASCADE). */
+		{
+			sqlite3_stmt *stmt_get_BLOBS_TABLE_size;
+			rc = sqlite3_prepare_v2(p_file_infos_arr->data[i]->db,
+				"SELECT MAX(Id) FROM " BLOBS_TABLE ";",
+				-1, &stmt_get_BLOBS_TABLE_size, NULL);
+			if (unlikely(rc != SQLITE_OK)) fatal_sqlite3_err(rc, __LINE__);
+			rc = sqlite3_step(stmt_get_BLOBS_TABLE_size);
+			if (unlikely(rc != SQLITE_ROW)) fatal_sqlite3_err(rc, __LINE__);
+			const int blobs_table_max_id = sqlite3_column_int(stmt_get_BLOBS_TABLE_size, 0);
+
+			sqlite3_stmt *stmt_retrieve_filename_last_digits; // This statement retrieves the last digit(s) from the Filename column of BLOBS_TABLE
+			rc = sqlite3_prepare_v2(p_file_infos_arr->data[i]->db,
+				"WITH split(word, str) AS ( SELECT '', (SELECT Filename FROM " BLOBS_TABLE " WHERE Id = ? ) || '.' "
+				"UNION ALL SELECT substr(str, 0, instr(str, '.')), substr(str, instr(str, '.')+1) FROM split WHERE str!='' ) "
+				"SELECT word FROM split WHERE word!='' ORDER BY LENGTH(str) LIMIT 1;",
+				-1, &stmt_retrieve_filename_last_digits, NULL);
+			if (unlikely(rc != SQLITE_OK)) fatal_sqlite3_err(rc, __LINE__);
+
+			sqlite3_stmt *stmt_delete_row_by_id; 
+			rc = sqlite3_prepare_v2(p_file_infos_arr->data[i]->db,
```

Comment:
> When we delete, we are also removing historical, are you sure it is not necessary to store information related to removed row?

If by this you mean keep the remove logs as archived, that's future work. We probably want to add a 2nd tier of "cold" storage, where the oldest logs get compressed (with higher compression ratio) and archived, instead of just deleted. Unless I missunderstood what you mean?

>When we remove in normal situation databases marks row as "invisible" and do not trough away it. Are you planning to run vacuum periodically in the table?

You are right about that. For the moment, autovacuum is enabled (see `PRAGMA auto_vacuum = INCREMENTAL;`), but it is not run. I am not sure yet if I will set it to autovacuum or do it manually. It needs to be handled somehow in either case. Also, there is the point of defragmentation, which in my understanding is not handled by autovacuuming (needs to be run manually periodically). I've added this in the TODO of the description.

>Considering that data will be discarded after usage, did you already think to use memory instead hard disk file?

I am not sure that's possible. We want to store logs on the disk so they can be queried later. And logs take up a lot of space, we wouldn't be able to just store them in RAM (if you are talking about just the metadata, I wanted to avoid corruption in case of a crash, so the last state can be recovered).

Reply:
> Unless I missunderstood what you mean?

You understood correctly. :)

Thank you to clarify, I always have concerns to remove data and we do not have conditions to recreate something, because the data was removed forever.